presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java

```diff
@@ -1718,6 +1718,150 @@ public void testRollupOverUnion()
+                "       (5, NULL, 105367.67, 1, 5)");
+    }
+
+    @Test
+    public void testGroupingInTableSubquery()
+            throws Exception
+    {
+        // In addition to testing grouping() in subqueries, the following tests also
+        // ensure correct behavior in the case of alternating GROUPING SETS and GROUP BY
+        // clauses in the same plan. This is significant because grouping() with GROUP BY
+        // works only with a special re-write that should not happen in the presence of
+        // GROUPING SETS.
+
+        // Inner query has a single GROUP BY and outer query has GROUPING SETS
+        assertQuery(
+            "SELECT orderkey, custkey, sum(agg_price) as outer_sum, grouping(orderkey, custkey), t.g " +
+                "FROM " +
+                "    (SELECT orderkey, custkey, sum(totalprice) as agg_price, grouping(custkey, orderkey) as g " +
+                "        FROM orders " +
+                "        GROUP BY orderkey, custkey " +
+                "        ORDER BY agg_price ASC " +
+                "        LIMIT 5) as t " +
+                "GROUP BY GROUPING SETS ((orderkey, custkey), t.g) " +
+                "ORDER BY outer_sum",
+            "VALUES (35271, 334, 874.89, 0, NULL), " +
+                "       (28647, 1351, 924.33, 0, NULL), " +
+                "       (58145, 862, 929.03, 0, NULL), " +
+                "       (8354, 634, 974.04, 0, NULL), " +
+                "       (37415, 301, 986.63, 0, NULL), " +
+                "       (NULL, NULL, 4688.92, 3, 0)");
+
+        // Inner query has GROUPING SETS and outer query has GROUP BY
+        assertQuery(
+            "SELECT orderkey, custkey, t.g, sum(agg_price) as outer_sum, grouping(orderkey, custkey) " +
+                "FROM " +
+                "    (SELECT orderkey, custkey, sum(totalprice) as agg_price, grouping(custkey, orderkey) as g " +
+                "     FROM orders " +
+                "     GROUP BY GROUPING SETS ((custkey), (orderkey)) " +
+                "     ORDER BY agg_price ASC " +
+                "     LIMIT 5) as t " +
+                "GROUP BY orderkey, custkey, t.g",
+            "VALUES (28647, NULL, 2, 924.33, 0), " +
+                "       (8354, NULL, 2, 974.04, 0), " +
+                "       (37415, NULL, 2, 986.63, 0), " +
+                "       (58145, NULL, 2, 929.03, 0), " +
+                "       (35271, NULL, 2, 874.89, 0)");
+
+        // Inner query has GROUPING SETS but no grouping and outer query has a simple GROUP BY
+        assertQuery(
+            "SELECT orderkey, custkey, sum(agg_price) as outer_sum, grouping(orderkey, custkey) " +
+                "FROM " +
+                "   (SELECT orderkey, custkey, sum(totalprice) as agg_price " +
+                "    FROM orders " +
+                "    GROUP BY GROUPING SETS ((custkey), (orderkey)) " +
+                "    ORDER BY agg_price ASC NULLS FIRST) as t " +
+                "GROUP BY orderkey, custkey " +
```

Comment:
Hehehe, this means the docs are incorrect. I'll update them, good catch. `grouping()` can be used with a simple `GROUP BY`.